### PR TITLE
Add support for UDTs to std.range.primitives.save

### PR DESCRIPTION
**Pull request by JakobOvrum**
When UDTs implement the `r.save` range primitive through subtyping a slice
with AliasThis, `save` would return the subtype instead of the derived type, thus failing to satisfy the `isForwardRange` concept.

I noticed this when working on std.exception.handle (see #2724 and #2916), which fails to propagate forward-rangeness and thus also bidirectionality when applied to slice types like strings (when `RangePrimitive.save` is *not* specified).

This seems like a sensible change to me; it allows std.exception.handle to remain elegant in implementation, which is good news for future range types wanting to use AliasThis as well. I don't think these kinds of problems will creep up on us in all kinds of places, but maybe I'm lacking imagination.

The specific trait that fails is `R r2 = r1.save;` which could technically also be implemented by adding an `opAssign` overload on the derived range type, but I think this would be an overly pedantic view of the `save` primitive. Maybe `save`'s definition needs to be slightly more strict so supertypes can forward it without going full-on substitution principle.

Ping @monarchdodra; I know you're wary of AliasThis problems. Note that in this case the AliasThis is indeed a simple lvalue field.

### CODE REVIEW

File: std/range/primitives.d [hunk 1]
```diff
@@ -1941,13 +1941,22 @@ arrays. Due to the fact that nonmember functions can be called with
...
+returned as their original type.
+ */
 @property T[] save(T)(T[] a) @safe pure nothrow
 {
     return a;
 }
 
+/// Ditto
+@property T save(T)(T a)
+    if((is(T == struct) || is(T == class)) && is(T : U[], U))
```
Comment (andralex) [inline std/range/primitives.d]: this won't work for classes

**Comment by JakobOvrum** [comment]
I created this as part of future PR that has tests, so I neglected creating tests. Will add tests now. I do believe it does work for classes.

File: std/range/primitives.d (hunk 1 above)
Comment (JakobOvrum) [inline std/range/primitives.d]: Added tests. It does work for classes, unless I'm missing something here.

**Comment by JakobOvrum** [comment]
Indeed it's not correct for classes, duh. Now only matches for structs, thanks @andralex!

**Comment by monarchdodra** [comment]
> Note that in this case the AliasThis is indeed a simple lvalue field.

That's just *one* of the many dangers of alias this ;)

--------

I'm unsure about this. If T is a UDT, then surely the author can choose whether or not save is implemented on it, and what it would mean to save said range, and the implications of doing so. "save" has a lot of implementation specific logic baked into it. I don't think it is wise for us to assume a generic way of implementing it.

Let's assume a standard wrapper this would impact and that also has this problem: `RefCounted`. The issue is that it models reference semantics, and you can't just assume that copy == save.

```D
    alias T = int[];
	alias RCT = RefCounted!T;
    auto a = RCT([1, 2, 3]);
    RCT b = a.save; //What does this do?
```
What's interesting here is that with the "old" behavior, b would actually be a brand new copy. But with the new behavior, it would just increment the ref count, and it certainly wouldn't actually save anything.

--------

> std.exception.handle

I'd strongly advocate that handle wrap *all* primitives, and use static ifs *inside* each implementation, to handle (or not) exceptions. In particular, save would become:
```D
        static if (isForwardRange!IRange)
        {
            @property typeof(this) save()
            {
                static if (primitivesToHandle & RangePrimitive.save)
                {
                    try
                    {
                        return typeof(this)(range.save);
                    }
                    catch(E exception)
                    {
                        return typeof(this)(handler(exception, this.range));
                    }
                }
                else
                {
                        return typeof(this)(range.save);
                }
            }
        }
```

`handle` is a range adaptor, and (AFAIK), no other range adaptor uses alias this.

Note: Makes me wish we had better tools to conditionally catch exceptions.

**Comment by JakobOvrum** [comment]
> I'm unsure about this. If T is a UDT, then surely the author can choose whether or not save is implemented on it, and what it would mean to save said range, and the implications of doing so. "save" has a lot of implementation specific logic baked into it. I don't think it is wise for us to assume a generic way of implementing it.

Agreed. The fact that I disabled this for classes is a good indicator; one would assume reference type structs would have exactly the same problem.

> I'd strongly advocate that handle wrap all primitives, and use static ifs inside each implementation, to handle (or not) exceptions. In particular, save would become:

Ya, the point was that with AliasThis we didn't need this repetition, and there would be no codegen required for a bunch of thin wrappers. But I agree it is the way to go.

> Note: Makes me wish we had better tools to conditionally catch exceptions.

Yeah, the awkwardness of try-catch tends to cause subtle problems from time to time.

I'll file a PR that refactors `handle`'s implementation into a traditional higher-order range.

**Comment by monarchdodra** [comment]
> Yeah, the awkwardness of try-catch tends to cause subtle problems from time to time.

"scope" with explicit exceptions would make this look ridiculously good:
```D
        static if (isForwardRange!IRange)
        {
            @property typeof(this) save()
            {
                static if (primitivesToHandle & RangePrimitive.save)
                     scope (E e)
                         return typeof(this)(handler(e, this.range));
                return typeof(this)(range.save);
            }
        }

**Pull request closed without merging**